packages/app-builder-lib/src/publish/KeygenPublisher.ts

```diff
@@ -0,0 +1,116 @@
+import { Arch, InvalidConfigurationError, log, isEmptyOrSpaces } from "builder-util"
+import { httpExecutor } from "builder-util/out/nodeHttpExecutor"
+import { ClientRequest, RequestOptions } from "http"
+import { HttpPublisher, PublishContext } from "electron-publish"
+import { KeygenOptions } from "builder-util-runtime/out/publishOptions"
+import { configureRequestOptions, HttpExecutor, parseJson } from "builder-util-runtime"
+import * as path from "path"
+export class KeygenPublisher extends HttpPublisher {
+  readonly providerName = "keygen"
+  readonly hostname = "api.keygen.sh"
+
+  private readonly info: KeygenOptions
+  private readonly auth: string
+  private readonly version: string
+  private readonly basePath: string
+
+  constructor(context: PublishContext, info: KeygenOptions, version: string) {
+    super(context)
+
+    const token = process.env.KEYGEN_TOKEN
+    if (isEmptyOrSpaces(token)) {
+      throw new InvalidConfigurationError(`Keygen token is not set using env "KEYGEN_TOKEN" (see https://www.electron.build/configuration/publish#KeygenOptions)`)
+    }
+
+    this.info = info
+    this.auth = `Bearer ${token.trim()}`
+    this.version = version
+    this.basePath = `/v1/accounts/${this.info.account}/releases`
+  }
+
+  protected doUpload(
+    fileName: string,
+    _arch: Arch,
+    dataLength: number,
+    requestProcessor: (request: ClientRequest, reject: (error: Error) => void) => void,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _file?: string
+  ): Promise<any> {
+    return HttpExecutor.retryOnServerError(async () => {
+      const { data, errors } = await this.upsertRelease(fileName, dataLength)
+      if (errors) {
+        throw new Error(`Keygen - Upserting release returned errors: ${JSON.stringify(errors)}`)
+      }
+      const releaseId = data?.id
```

Comment:
Note: the release ID will always be present, given there are no `errors` (which is already handled above), so checking if it exists here isn't strictly needed.

Reply:
More so personal preference, I just strongly dislike force casting or force unwrapping a variable